devtools/etdump/buffer_data_sink.h

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <executorch/devtools/etdump/data_sink_base.h>
+#include <executorch/runtime/core/exec_aten/exec_aten.h>
+#include <executorch/runtime/core/span.h>
+
+namespace executorch {
+namespace etdump {
+
+/**
+ * BufferDataSink is a concrete implementation of the DataSinkBase class,
+ * designed to store debug data in a pre-allocated, user-owned buffer. This
+ * class provides methods to write raw data and tensor data into the buffer,
+ * ensuring proper alignment and managing padding as needed. It is the standard
+ * DataSink used by ETDumpGen.
+ */
+class BufferDataSink : public DataSinkBase {
+ public:
+  /**
+   * Constructs a BufferDataSink with a given buffer.
+   *
+   * @param[in] buffer A Span object representing the buffer where data will be
+   * stored.
+   */
+  explicit BufferDataSink(::executorch::runtime::Span<uint8_t> buffer)
+      : debug_buffer_(buffer), offset_(0), alighment_(64) {}
```

Comment:
May as well add an alignment param. And please document that it must be a power of 2
```suggestion
  explicit BufferDataSink(::executorch::runtime::Span<uint8_t> buffer, size_t alignment = 64)
      : debug_buffer_(buffer), offset_(0), alignment_(alignment) {}
```